HDDS-5449 Recon namespace summary 'du' information should return replicated size of a key

yuangu002 opened the pull request:
## What changes were proposed in this pull request?

In addition to actual key size, Recon 'du' should also return the replicated size of a key.
For example, for a 3 way replicated key of size = 1MB, Recon 'du' should return '1MB' as the size of the key, and '3MB' as the replicated size if the key is fully replicated.

[Refactoring]
1. DU response now shows a "structure"
2. DU key response shouldn't have subpaths

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5449

## How was this patch tested?

Since this is an improvement/refactoring feature on top of HDDS-5378, no additional unit test is implemented. I tested remotely on the cluster endpoint with the following configs:
![Screen Shot 2021-08-03 at 1 46 36 PM](https://user-images.githubusercontent.com/53324985/128062467-ff763bc6-07a0-4a5a-9eab-857b8643523a.png)
All keys are three-way replicated except `key3`

[Root DU]
http://localhost:9888/api/v1/namespace/du?path=/&replica=true
![root du -rep](https://user-images.githubusercontent.com/53324985/128065104-1e4fdbdd-9b84-4c75-9b14-cf70fb7ea5c7.png)

[Volume DU]
http://localhost:9888/api/v1/namespace/du?path=/vol&replica=true
![vol du -rep](https://user-images.githubusercontent.com/53324985/128065167-758ffd63-e89f-4120-a03b-3f7acea11715.png)

[Bucket DU] `&files=true` enabled
http://localhost:9888/api/v1/namespace/du?path=/vol/bucket&files=true&replica=true
![bucket du -rep](https://user-images.githubusercontent.com/53324985/128065213-9049b519-27a6-443c-b387-450b9b9f2993.png)

[Directory DU] `&files=true` enabled
http://localhost:9888/api/v1/namespace/du?path=/vol/bucket/dir1&files=true&replica=true
![dir1 du -rep](https://user-images.githubusercontent.com/53324985/128065242-e6b9efbd-532d-42e7-b2ac-9d913b8e8cff.png)
(Key3 is one-way replicated.)

[Nested Directory DU]
http://localhost:9888/api/v1/namespace/du?path=/vol/bucket/dir1/dir2&files=true&replica=true
![dir2 du -rep](https://user-images.githubusercontent.com/53324985/128065272-15b0af28-b79e-4f98-9666-f4c22ec9bf53.png)

[Key DU]
http://localhost:9888/api/v1/namespace/du?path=/vol/bucket/dir1/dir2/dir3/key6&files=true&replica=true
![key du -rep](https://user-images.githubusercontent.com/53324985/128065331-8f0f3023-f50a-47a7-84f1-65774524708d.png)

yuangu002 commented:
cc @smengcl @avijayanhwx. Please review

avijayanhwx commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -62,17 +72,25 @@
     this.reconNamespaceSummaryManager = namespaceSummaryManager;
     this.omMetadataManager = omMetadataManager;
+    this.containerManager =
+            (ReconContainerManager) reconSCM.getContainerManager();
```
We don't need the cast to RCM here. We should be OK with the return type being 'ContainerManagerV2' since we are using only interface method.

avijayanhwx commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -163,12 +181,19 @@ public Response getBasicInfo(
+                               @DefaultValue("false")
+                               @QueryParam("files") boolean listFile,
+                               @DefaultValue("false")
+                               @QueryParam("replica") boolean withReplica)
```
Can we set Default value for replica?

yuangu002 replied:
shouldn't `@DefaultValue("false")` suffice?

avijayanhwx commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -267,26 +368,69 @@ public Response getDiskUsage(@QueryParam("path") String path)
+        duResponse.setSizeWithReplica(keySizeWithReplica);
+      }
+      // give an empty list to avoid NPE
+      duResponse.setDuData(new ArrayList<>());
```
We could init the duData list to empty list in the DuResponse constructor.

avijayanhwx commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -724,6 +868,135 @@ private long getTotalSize(long objectId) throws IOException {
+            iterator = keyTable.iterator();
+
+    while (iterator.hasNext()) {
+      Table.KeyValue<String, OmKeyInfo> kv = iterator.next();
```
Can we use the iterator.seek() API that let's us jump to the first key with that the volume name in prefix here?

avijayanhwx commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -724,6 +868,135 @@ private long getTotalSize(long objectId) throws IOException {
+            iterator = keyTable.iterator();
+
+    while (iterator.hasNext()) {
+      Table.KeyValue<String, OmKeyInfo> kv = iterator.next();
```
Same as above.

avijayanhwx commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -724,6 +868,135 @@ private long getTotalSize(long objectId) throws IOException {
+    TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+            iterator = keyTable.iterator();
+
+    while (iterator.hasNext()) {
```
Same as above.

avijayanhwx commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryEndpoint.java:
```diff
@@ -299,13 +311,15 @@ public void testDiskUsage() throws Exception {
     // key level DU
-    Response keyResponse = nsSummaryEndpoint.getDiskUsage(KEY_PATH);
+    Response keyResponse = nsSummaryEndpoint.getDiskUsage(KEY_PATH,
+            false, false);
```
Can we add tests for replica size? There are a few cases that need to be verified (key spanning into multiple blocks, container replica count < desired replica count etc)

avijayanhwx reviewed (commented):
Thanks for working on this @yuangu002. A few comments inline.

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -267,26 +368,69 @@ public Response getDiskUsage(@QueryParam("path") String path)
+
+      // handle direct keys under directory
+      if (listFile || withReplica) {
+        List<OmKeyInfo> directKeys = listDirectKeys(dirObjectId);
```
Same here. We can calculate the replica directly here without gathering all `OmKeyInfo`s

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -267,26 +368,69 @@ public Response getDiskUsage(@QueryParam("path") String path)
+      if (listFile || withReplica) {
+        List<OmKeyInfo> directKeys = listDirectKeys(dirObjectId);
+
+        for (OmKeyInfo directKey: directKeys) {
```
```suggestion
        for (OmKeyInfo directKey : directKeys) {
```

yuangu002 commented:
[Fix]
1. Addressed comments.
2. Optimized listKeys by seek() and DFS. No full iteration on FileTable except root level
3. Added a unit test for the case that a key spanning across multiple blocks and some of containers are under/over-replicated

[Future works]
1. https://github.com/yuangu002/ozone/blob/HDDS-5449/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java#L228
2. https://github.com/yuangu002/ozone/blob/HDDS-5449/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java#L907
3. Change `numOfFiles` in `NSSummary` from `int` to `long`

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -991,7 +947,7 @@ private long getKeySizeWithReplication(OmKeyInfo keyInfo) {
         du += blockSize;
       } catch (ContainerNotFoundException cnfe) {
-        LOG.warn("Cannot find container");
+        LOG.debug("Cannot find container {}", block.getContainerID(), cnfe);
```
If this is a rare exception, we should leave this at `warn`.

smengcl commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryEndpoint.java:
```diff
@@ -51,16 +57,13 @@
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.List;
+import java.util.*;
```
Let's change this back.

You can disable intellij's wildcard imports feature: https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcards-for-class

smengcl commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryEndpoint.java:
```diff
@@ -18,13 +18,19 @@
 import org.apache.hadoop.ozone.om.OmMetadataManagerImpl;
-import org.apache.hadoop.ozone.om.helpers.OmBucketInfo;
-import org.apache.hadoop.ozone.om.helpers.OmVolumeArgs;
+import org.apache.hadoop.ozone.om.helpers.*;
```
This as well.

smengcl reviewed (approved):
Latest changes lgtm.

Pending CI.

smengcl commented:
Merged. Thanks @yuangu002 for the patch. Thanks @avijayanhwx for the review.
